docs/design/sql/10-distributed-scan.md

```diff
@@ -0,0 +1,164 @@
+created and destroyed. This section describes the behavior of scan operators in the case of concurrent cluster reconfiguration.
+
+### 4.1 Partition Migration
+
+If the query is executed concurrently with partition migration, it is possible that some partitions will be scanned several 
+times, or missed completely. To avoid that we need a mechanics to ensure that the engine observed every partition exactly
+once.  
+
+When the plan is created, we create the so-called **partition map**, that maps members to partitions. The partition map is saved
+in the plan. If the partition distribution changes due to member join/leave, the plan is invalidated.  
+
+When the query is about to start, we explicitly assign partitions to every participant. For example:
+```
+member1 -> [1, 2]
+member2 -> [3, 4]
+```  
+
+When the scan operation is started on the member, we first check whether it contains expected partitions. If the expected and 
+actually owned partitions do not match, an exception is thrown. For example, if the member is expected to have partitions 
+`[1, 2]`, then:
+```
+[1] - error, because we may miss the partition [2]
+[1, 2] - ok
+[1, 2, 3] - error, because we may return duplicates for the partition [3]
+```
+
+The check before the query start is not sufficient to guarantee the correctness of the results. The partition may migrate to the 
+member during execution of the query, leading to duplicates. Conversely, the partition may be removed from the member 
+concurrently, leading to missed entries. To avoid inconsistent results we **re-check** the partition distribution before any
+result is returned from the scan operator. 
+
+Partition check logic depends on the operator type. For the direct map scan, we use the migration stamps 
+(`MapService.validateMigrationStamp`). For the index scans, we use index partition stamps 
+(`InternalIndex.validatePartitionStamp`).
+
+A better solution would hide the partitioning problems from the user completely. This, however, is difficult to achieve 
+in practice. To avoid duplicates and missed entries we would have to track which part of the operator input is processed, and 
+re-schedule scans to other members at runtime. Therefore, the current design ensures that the user does not see an inconsistent
+result, but forces the user to re-run the query manually if the result correctness cannot be guaranteed by the engine. In the
+long term we would like to avoid, or at least minimize, the number of cases when the error is thrown.
+
+### 4.2 Concurrent Schema Changes
+
+It may happen, that the referenced object (`IMap`, index) that was present during the query planning, no longer exists on the 
+local member. For example, because the map has been destroyed concurrently. Therefore, when the scan operation is about to
+start, the existence of the required objects is checked. If the required object is not found, an exception is thrown, and the
```

Comment:
What happens if the object is deleted during the scan execution?

Reply:
Clarified in the commit: https://github.com/hazelcast/hazelcast/pull/17659/commits/d40f45d8a828f99617ed0ff334d338754592eb73